blueprints/user.permissions.md

```diff
@@ -0,0 +1,286 @@
+is as follows:
+
+```typescript
+interface Role {
+	capabilities: Array<string> | null;
+	description: string | null;
+	id?: number | null;
+	name: string | null;
+	privLevel: number | null;
+}
+```
+
+The proposed new model is shown below.
+
+```typescript
+interface Role {
+	description: string;
+	lastUpdated?: Date;
+	name: string;
+	permissions: Array<string>;
+}
+```
+
+Also, because the ID is no longer exposed via the API, the model for users will
+need to change to only reference Role name, not ID. So it'll change from...
+
+```typescript
+interface User {
+	addressLine1: string | null;
+	addressLine2: string | null;
+	city: string | null;
+	company: string | null;
+	confirmLocalPassword?: string | null;
+	country: string | null;
+	email: string | null;
+	fullName: string | null;
+	gid: number | null;
+	id?: number | null;
+	lastUpdated?: string | null;
+	localPassword?: string | null;
+	newUser: boolean | null;
+	phoneNumber: string | null;
+	postalCode: string | null;
+	publicSSHKey: string | null;
+	registrationSent: string | null;
+	role: number | null;
+	roleName?: string | null;
+	stateOrProvince: string | null;
+	tenant?: string | null;
+	tenantID: string | null;
+	uid: number | null;
+	username: string | null;
+}
+```
+
+... to ...
+
+```typescript
+interface User {
+	addressLine1: string | null;
+	addressLine2: string | null;
+	city: string | null;
+	company: string | null;
+	confirmLocalPassword?: string | null;
+	country: string | null;
+	email: string | null;
+	fullName: string;
+	gid: number | null;
+	id?: number;
+	lastUpdated?: Date;
+	localPassword?: string | null;
+	newUser: boolean;
+	phoneNumber: string | null;
+	postalCode: string | null;
+	publicSSHKey: string | null;
+	registrationSent: string | null;
+	role: string;
+	stateOrProvince: string | null;
+	tenant?: string;
+	tenantID: number;
+	uid: number | null;
+	username: string;
+}
+```
+
+... replacing the required `role` field that was a Role ID with one that is
+the Role's name (and therefore removing `roleName` as it's now redundant).
+
+(Note that `lastUpdated` changed from `string | null` to `Date`, which
+represents the transition to using proper RFC3339 formatting, and also some
+fields which cannot be `null` for a valid user are no longer `| null`. Those
+aren't changes necessary to implement Permissions, but can be done easily at
+the same time as necessary changes to User.)
+
+## Component Impact
+This change primarily concerns Traffic Ops and Traffic Portal, though any
+client of the API will need to be aware of the changes to its authorization
+system.
+
+### Traffic Portal Impact
+There is a "Capabilities" view currently in Traffic Portal that provides a
+table of user Permissions, though it is not linked to the side navigation bar.
+This should be removed.
+
+The Roles' details pages will need to be augmented with controls for adding and
+removing Permissions. The Permissions available for adding to a user should
```

Comment:
> The Permissions available for adding to a user

this sounds like you add perms to a user. is that true? or do you add perms to a role (which belongs to a user0?

Reply:
yeah, that should say Role.